### PR TITLE
Python: Add tests for improved Provisioning Error Messages

### PR DESCRIPTION
**Pull request by KGB33**
Follow up to PR #3825 

The test names could probably use some work, let me know if there is a standard format you'd like to use and I'll fix them up. 
Maybe something like `test_raises_EXCEPTION_when_CONDITION`?

Also If I missed any test paths let me know!

### CODE REVIEW

**Comment by KGB33** [comment]
Currently working on getting `./hack/make sdk:python:test` to pass.

**Comment by KGB33** [comment]
Got the tests to pass `./hack/make sdk:python:test`. 

The three tests that don't start docker/dagger just mock out the `ImageRef` object.

The because the dagger engine is started in the last test (`test_docker_engine_is_not_running_cached_dagger_engine_exists`) I marked it as skipped when `dagger.Config().host.scheme != "docker-image"`.

**Comment by helderco** [comment]
Thanks for taking care of this so quickly, I'll review tomorrow 🙂

File: sdk/python/tests/connectors/test_docker.py [hunk 1]
```diff
@@ -1,25 +1,50 @@
...
-)
-async def test_docker_image_provision(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+@pytest.fixture
+def cache_dir(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+    """Creates a temp cache_dir for testing & sets XDG_CACHE_HOME."""
     cache_dir = tmp_path / "dagger"
     cache_dir.mkdir()
     monkeypatch.setenv("XDG_CACHE_HOME", str(tmp_path))
+    yield cache_dir
+    shutil.rmtree(str(cache_dir))
```
Comment (helderco) [inline sdk/python/tests/connectors/test_docker.py:28]: `tmp_path` should be cleaned after each test so this isn't needed.

File: sdk/python/tests/connectors/test_docker.py [hunk 2]
```diff
@@ -45,3 +70,90 @@ async def connect_once():
...
+    """
+    When the docker cli is not installed ensure that the `FileNotFoundError` returned by
+    `subprocess.run` is wrapped by a `docker.ProvisionError` stating that
+    the command is not found.
+    """
+
+    def pached_subprocess_run(*args, **kwargs):
+        raise FileNotFoundError()
+
+    monkeypatch.setattr(docker.subprocess, "run", pached_subprocess_run)
```
Comment (helderco) [inline sdk/python/tests/connectors/test_docker.py:87]: Using [pytest-subprocess](https://pytest-subprocess.readthedocs.io/en/latest/usage.html#basic-usage) can make this easier.

File: sdk/python/tests/connectors/test_docker.py [hunk 3]
```diff
@@ -45,3 +70,90 @@ async def connect_once():
         next(files)
     # assert the garbage was cleaned up
     assert not garbage_path.exists()
+
+
+def test_docker_cli_is_not_installed(
```
Comment (helderco) [inline sdk/python/tests/connectors/test_docker.py:75]: These tests are marked with `anyio` from the module (i.e., _async_). Sync tests should remove the mark or go into a separate `test_docker_sync.py` file.

File: sdk/python/tests/connectors/test_docker.py [hunk 4]
```diff
@@ -45,3 +70,90 @@ async def connect_once():
...
+    ensure that the `ValueError` is wrapped in a `ProvisionError`
+    and a more helpful error message is printed.
+    """
+    # Create the cached 'dagger engine'
+    cfg = Config(log_output=log_output)
+    async with dagger.Connection(cfg):
+        pass
+
+    # Mock DOCKER_HOST to make it seem like docker isn't running.
+    monkeypatch.setenv("DOCKER_HOST", "127.1.2.3")
```
Comment (helderco) [inline sdk/python/tests/connectors/test_docker.py:156]: `DOCKER_HOST` should have a scheme, so this should be `http://127.1.2.3`.

**Review by helderco — COMMENTED** [review]
I made a few quick comments but I want to wait for https://github.com/dagger/dagger/pull/3787 because it'll split the _docker_ connector into an additional _bin_ connector.

Then I'll review this again.

Thanks! 🙏

File: sdk/python/tests/connectors/test_docker.py (hunk 3 above)
Comment (KGB33) [reply sdk/python/tests/connectors/test_docker.py:75]: I opted to keep all the tests in the same file, it seemed less confusing with the `sync`/`async` apis.

Comment (helderco) [reply sdk/python/tests/connectors/test_docker.py:75]: What I mean is that there's module wide [marks](https://docs.pytest.org/en/7.2.x/how-to/mark.html) in this file for `anyio` and `slow`:

```python
pytestmark = [
    pytest.mark.anyio,
    pytest.mark.slow,
]
```

Tests should be marked appropriately, so either put them in each test or split in different files.

**Comment by helderco** [comment]
@KGB33 https://github.com/dagger/dagger/pull/3787 has been merged so please rebase and update accordingly, I'll review on monday. 🙏

File: sdk/python/tests/connectors/test_docker.py (hunk 3 above)
Comment (KGB33) [reply sdk/python/tests/connectors/test_docker.py]: Sorry, I should have clarified (or waited till I pushed the commit), I kept all the tests in the same file & marked each test appropriately.

**Review by helderco — APPROVED** [review]
It's been crazy lately with braking changes and not being able to test locally properly. The provisioning stuff is still in flux and will suffer major changes soon. I'm going to merge this as everything else has been rebased.

**Pull request merged**